### DISCUSSION TITLE
Configurable delimiter for JSON export

### DISCUSSION THREAD

**Discussion by adanielyan** [tolgee/tolgee-platform#1462] [Ideas]
For some of my translations I use the actual text as key. It allows me to fallback to the key if there's no translation available (I use i18next which provides that functionally). As a result some of my keys have dots in them which are not meant to be a delimiter but are indicating the end of the sentence. The toglee platform handles those keys without any problem. However, when I try to export the translation to json through API I get nested objects for every dot in a key. I think to avoid this behavior it would be great if it was possible to configure the delimiter symbol (or combination of symbols, e.g. ```->```) in the project settings.

**Comment by JanCizmar** [comment]
Hi there. When using the [`/v2/projects/export`](https://tolgee.io/api#tag/Export/operation/export_1), you can provide the `splitByScopeDelimiter `. However, this parameter will be renamed to `structureDelimiter` in Tolgee Platform v3, so keep that in mind.

**Reply by adanielyan** [reply]
Thank you for your advice and for creating a great open source product!

However, I need to control the delimiter on this API endpoint: ```/v2/projects/translations/{language}``` and the query param you mentioned doesn't work for it. Is there a plan to add support for ```structureDelimiter``` to this endpoint too?

I also noticed that the ```/v2/projects/translations?languages={language}``` returns json translations without breaking the keys by delimiters. It does what I need but at the same time returns a lot of information that I don't need, making the response size way too big.

I think this behavior is inconsistent - returning nested objects for every delimiter on one API endpoint and keeping the keys as is on another. Having consistency in how the keys are treated or at least what query params are supported on similar API endpoints would be very helpful.

**Reply by JanCizmar** [reply]
Can you please describe your set-up and use case?

**Reply by adanielyan** [reply]
Sure. I use the translations in a mobile app that can work both with and without internet connection. The app will be shipped with initial translation JSON files. However, when internet connection is available, the app will pull the latest translations from API and store them locally for offline use. The translations are updated on a regular basis, so it is important to have the app pull and store the new translations whenever internet is available. It is critical that the translations get updated every time there is an internet connection, without having to create and ship another build of the app.

**Reply by JanCizmar** [reply]
OK, It is valid use-case. I will add the parameter today. :)

**Reply by adanielyan** [reply]
Thank you so much, it would be great!

**Reply by JanCizmar** [reply]
It's fixed in this PR. https://github.com/tolgee/tolgee-platform/pull/1463

When CRs are complete, I will cherrypick it also to main, so you can get it asap.

**Reply by adanielyan** [reply]
Thank you, I really appreciate such a quick turnaround!

**Reply by adanielyan** [reply]
I finally had a chance to test the fix that you implemented for translations API. It now returns all of the translations in a flat JSON object, as I needed, which is great. 

However none of the delimiter parameters work when provided `(splitByScope, splitByScopeDelimiter, splitByScopeDepth)`. I thought that the delimiter parameters would still work if provided. Isn't it the case? In some cases I have a namespace which is the first string in the key separated by dot (e.g. `common.Please provide your email`). In this case I would use `splitByScope=true&splitByScopeDelimiter=.&splitByScopeDepth=1`. This is not a dealbreaker and I can process that part in my code, but wanted to check with you first.

**Reply by JanCizmar** [reply]
You are still confusing the translation endpoint with the export endpoint. All the parameters you described are defined ONLY for the export endpoint. For the translations endpoint I've added only the `structureDelimiter` parameter. Anyway, i just checked it and it really returns a flat JSON by default, which is a bug. It should do that only when structureDelimiter is set to empty string. I am fixing it now.

**Reply by adanielyan** [reply]
I see. Would it be possible to add `structureDelimiterDepth` parameter for consistency with the export endpoint?

**Reply by JanCizmar** [reply]
Since tolgee-platform v3 has native support for namespaces, it would be better to you start using the [v3.0.0-alpha.3](https://hub.docker.com/layers/tolgee/tolgee/v3.0.0-alpha.3/images/sha256-83e336f8b9bf4e2746292fea99340367a34aca89a7c83779ac26e7ffcbcacf4d?context=explore). All the parameters are removed in v3 and replaced only with the structureDelimiter.

**Reply by adanielyan** [reply]
OK, using namespaces makes a lot of sense. Is `structureDelimiter` parameter available on `translations` endpoint too?

**Reply by JanCizmar** [reply]
Yes, that's available in both. You can check it by accessing the swagger definition which is accessible as `/swagger-ui/index.html` on your instance.

**Reply by adanielyan** [reply]
Great, I tested it and it worked. Now when I create a couple of namespaces and create a few translations there, the `/translation/{language}` endpoint only returns the translations that have no namespace, but not the newly created ones under the new namespaces. When I add param `ns=comon` I get the translations under that namespace. But if I add `ns=common,help` or `ns=common&ns=help` where `help` is another namespace I have, I get an empty object. Does this mean that "namespaced" translations cannot be retrieved together with the translations that don't have a namespace assigned? Also, does it mean I can only retrieve translations for one namespace at a time?

**Reply by JanCizmar** [reply]
> Does this mean that "namespaced" translations cannot be retrieved together with the translations that don't have a namespace assigned?
Yes

> Also, does it mean I can only retrieve translations for one namespace at a time?
Yes

The endpoint is actually there for our integrations, where we fetch namespace data only when requested. That's why we don't need to fetch multiple namespaces at the same time.

**Reply by adanielyan** [reply]
OK, thanks for clarification!

**Comment by adanielyan** [comment]
It may be a stupid question, but how do I update my docker container to run the latest version of tolgee-platform?

**Reply by JanCizmar** [reply]
I think you should open a new thread for this. 👍🏻
